spectator/config.go

```diff
@@ -1,29 +1,94 @@
 package spectator
 
 import (
+	"fmt"
 	"github.com/Netflix/spectator-go/spectator/logger"
 	"github.com/Netflix/spectator-go/spectator/writer"
 	"os"
 )
 
 // Config represents the Registry's configuration.
 type Config struct {
-	Location   string
-	CommonTags map[string]string
-	Log        logger.Logger
+	location   string
+	commonTags map[string]string
+	log        logger.Logger
 }
 
-// GetLocation Checks the location provided in the config and the environment variable SPECTATOR_OUTPUT_LOCATION.
-// If neither are valid, returns the default UDP address.
-func (c *Config) GetLocation() string {
-	configValue := c.Location
-	envValue := os.Getenv("SPECTATOR_OUTPUT_LOCATION")
-
-	if writer.ValidOutputLocation(configValue) {
-		return configValue
-	} else if writer.ValidOutputLocation(envValue) {
-		return envValue
-	} else {
-		return "udp://127.0.0.1:1234"
+// NewConfig creates a new configuration with the provided location, common tags, and logger. All fields are optional.
+// Possible values for location are:
+//   - `none`: Configures a no-op writer that does nothing. Can be used to disable metrics collection.
+//   - `stdout`: Writes metrics to stdout.
+//   - `stderr`: Writes metrics to stderr.
+//   - `memory`: Writes metrics to memory. Useful for testing.
+//   - `file:///path/to/file`: Writes metrics to a file.
+//   - `unix:///path/to/socket`: Writes metrics to a Unix domain socket.
+//   - `udp://host:port`: Writes metrics to a UDP socket.
+//
+// If location is not provided, the default value `udp://127.0.0.1:1234` will be used
+func NewConfig(
+	location string, // defaults to `udp://127.0.0.1:1234`
+	commonTags map[string]string, // defaults to empty map
+	log logger.Logger, // defaults to default logger
+) (*Config, error) {
+	location, err := calculateLocation(location)
+	if err != nil {
+		return nil, err
 	}
+
+	mergedTags := calculateCommonTags(commonTags)
+
+	lg := calculateLogger(log)
+
+	return &Config{
+		location:   location,
+		commonTags: mergedTags,
+		log:        lg,
+	}, nil
+}
+
+func calculateLogger(log logger.Logger) logger.Logger {
+	lg := log
+	if log == nil {
+		lg = logger.NewDefaultLogger()
+	}
+	return lg
+}
+
+func calculateCommonTags(commonTags map[string]string) map[string]string {
+	mergedTags := make(map[string]string)
+
+	for k, v := range commonTags {
```

Comment:
FYI there is a maps package that provides some niceties for cloning/copying values, though it doesn't filter the way you're doing here  https://pkg.go.dev/golang.org/x/exp/maps